File: fsharp/BufferHash.fs
namespace Benchmarks

open System

module HashAlgorithms =
    module SHA256Like =
        let private initialHashes = [|
            0x6a09e667u
            0xbb67ae85u
            0x3c6ef372u
            0xa54ff53au
            0x510e527fu
            0x9b05688cu
            0x1f83d9abu
            0x5be0cd19u
        |]

        let compute (data: byte[]) =
            let hashes = Array.copy initialHashes

            data
            |> Array.iteri (fun i b ->
                let hashIdx = i % 8
                let b = uint32 b

                let hash = 
                    hashes.[hashIdx]
                    |> fun h -> ((h <<< 5) + h) + b
                    |> fun h -> h &&& 0xFFFFFFFFu
                    |> fun h -> (h + (h <<< 10)) ^^^ (h >>> 6)
                    |> fun h -> h &&& 0xFFFFFFFFu

                hashes.[hashIdx] <- hash)

            let result = Array.zeroCreate<byte> 32

            for i = 0 to 7 do
                let hash = hashes.[i]
                result.[i * 4] <- byte ((hash >>> 24) &&& 0xFFu)
                result.[i * 4 + 1] <- byte ((hash >>> 16) &&& 0xFFu)
                result.[i * 4 + 2] <- byte ((hash >>> 8) &&& 0xFFu)
                result.[i * 4 + 3] <- byte (hash &&& 0xFFu)

            (uint32 result.[0]) |||
            (uint32 result.[1] <<< 8) |||
            (uint32 result.[2] <<< 16) |||
            (uint32 result.[3] <<< 24)

    module CRC32 =
        let compute (data: byte[]) =
            let mutable crc = 0xFFFFFFFFu

            for b in data do
                crc <- crc ^^^ (uint32 b)

                for _ = 1 to 8 do
                    if (crc &&& 1u) <> 0u then
                        crc <- (crc >>> 1) ^^^ 0xEDB88320u
                    else
                        crc <- crc >>> 1

            crc ^^^ 0xFFFFFFFFu

[<AbstractClass>]
type BufferHashBenchmark() =
    inherit Benchmark()

    let mutable data = Array.empty<byte>
    let mutable result = 0u

    member _.Data = data
    member _.UpdateResult value = result <- result + value

    override this.Prepare() =
        let n = int (this.ConfigVal("size"))
        data <- Array.init n (fun _ -> byte (Helper.NextInt(256)))
        result <- 0u

    override this.Checksum = result

    abstract member ComputeHash: byte[] -> uint32

    override this.Run(_: int64) =
        let hashValue = this.ComputeHash(data)
        this.UpdateResult hashValue

type BufferHashSHA256() =
    inherit BufferHashBenchmark()

    override _.ComputeHash(data: byte[]) = HashAlgorithms.SHA256Like.compute data

type BufferHashCRC32() =
    inherit BufferHashBenchmark()

    override _.ComputeHash(data: byte[]) = HashAlgorithms.CRC32.compute data

File: julia/src/benchmarks/BufferHash.jl
using ..BenchmarkFramework

abstract type AbstractBufferHashBenchmark <: AbstractBenchmark end

mutable struct BufferHashBenchmark <: AbstractBufferHashBenchmark
    size::Int64
    data::Vector{UInt8}
    result::UInt32

    function BufferHashBenchmark()
        size_val = Helper.config_i64("BufferHashBenchmark", "size")
        new(size_val, Vector{UInt8}(undef, size_val), UInt32(0))
    end
end

name(b::BufferHashBenchmark)::String = "BufferHashBenchmark"

function prepare(b::BufferHashBenchmark)
    for i in 1:length(b.data)
        b.data[i] = UInt8(Helper.next_int(256))
    end
end

function test(b::BufferHashBenchmark)::UInt32

    error("Abstract method 'test' not implemented")
end

function run(b::BufferHashBenchmark, iteration_id::Int64)
    hash_val = test(b)
    b.result = (b.result + hash_val) & 0xffffffff
end

function checksum(b::BufferHashBenchmark)::UInt32
    return b.result
end

mutable struct BufferHashSHA256 <: AbstractBufferHashBenchmark
    size::Int64
    data::Vector{UInt8}
    result::UInt32

    function BufferHashSHA256()
        size_val = Helper.config_i64("BufferHashSHA256", "size")
        new(size_val, Vector{UInt8}(undef, size_val), UInt32(0))
    end
end

name(b::BufferHashSHA256)::String = "BufferHashSHA256"

function prepare(b::BufferHashSHA256)
    for i in 1:length(b.data)
        b.data[i] = UInt8(Helper.next_int(256))
    end
end

module SimpleSHA256
    const INITIAL_HASHES = UInt32[
        0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a,
        0x510e527f, 0x9b05688c, 0x1f83d9ab, 0x5be0cd19
    ]

    function digest(data::Vector{UInt8})::Vector{UInt8}
        hashes = copy(INITIAL_HASHES)

        for (i, byte_val) in enumerate(data)
            byte = UInt32(byte_val)
            hash_idx = (i - 1) % 8 + 1  

            hash = hashes[hash_idx]
            hash = ((hash << 5) + hash) + byte

            hash = (hash + (hash << 10)) ⊻ (hash >> 6)

            hashes[hash_idx] = hash
        end

        result = Vector{UInt8}(undef, 32)
        for i in 1:8
            hash = hashes[i]

            result[(i-1)*4 + 1] = UInt8((hash >> 24) & 0xff)  
            result[(i-1)*4 + 2] = UInt8((hash >> 16) & 0xff)
            result[(i-1)*4 + 3] = UInt8((hash >> 8) & 0xff)
            result[(i-1)*4 + 4] = UInt8(hash & 0xff)          
        end

        return result
    end
end

function test(b::BufferHashSHA256)::UInt32
    bytes = SimpleSHA256.digest(b.data)

    return reinterpret(UInt32, bytes[1:4])[1]  

    b0 = UInt32(bytes[1])  
    b1 = UInt32(bytes[2])
    b2 = UInt32(bytes[3])
    b3 = UInt32(bytes[4])  

    return (b3 << 24) | (b2 << 16) | (b1 << 8) | b0
end

function run(b::BufferHashSHA256, iteration_id::Int64)
    hash_val = test(b)
    b.result = (b.result + hash_val) & 0xffffffff
end

function checksum(b::BufferHashSHA256)::UInt32
    return b.result
end

mutable struct BufferHashCRC32 <: AbstractBufferHashBenchmark
    size::Int64
    data::Vector{UInt8}
    result::UInt32

    function BufferHashCRC32()
        size_val = Helper.config_i64("BufferHashCRC32", "size")
        new(size_val, Vector{UInt8}(undef, size_val), UInt32(0))
    end
end

name(b::BufferHashCRC32)::String = "BufferHashCRC32"

function prepare(b::BufferHashCRC32)
    for i in 1:length(b.data)
        b.data[i] = UInt8(Helper.next_int(256))
    end
end

function crc32_compute(data::Vector{UInt8})::UInt32
    crc = 0xffffffff

    for byte in data
        crc = crc ⊻ UInt32(byte)
        for _ in 1:8
            if (crc & 1) != 0
                crc = (crc >> 1) ⊻ 0xedb88320
            else
                crc = crc >> 1
            end
        end
    end

    return crc ⊻ 0xffffffff
end

function test(b::BufferHashCRC32)::UInt32
    return crc32_compute(b.data)
end

function run(b::BufferHashCRC32, iteration_id::Int64)
    hash_val = test(b)
    b.result = (b.result + hash_val) & 0xffffffff
end

function checksum(b::BufferHashCRC32)::UInt32
    return b.result
end